engine/src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala

```diff
@@ -181,15 +180,38 @@ class SlickDataAccess(databaseConfig: Config) extends DataAccess {
     val actionThreadPoolSize = databaseConfig.getIntOr("actionThreadPoolSize", dbNumThreads) min dbMaximumPoolSize
     Executors.newFixedThreadPool(actionThreadPoolSize)
   }
+
   private val actionExecutionContext: ExecutionContext = ExecutionContext.fromExecutor(
-    actionThreadPool, database.executor.executionContext.reportFailure)
+    actionThreadPool, database.executor.executionContext.reportFailure
+  )
+
+  private lazy val useSlickUpserts = dataAccess.driver.capabilities.contains(JdbcProfile.capabilities.insertOrUpdate)
 
   override def close(): Unit = {
     actionThreadPool.shutdown()
     database.close()
   }
 
-  // Run action with an outer transaction
+  private def runTransactionWithRollbackRetry[R](action: DBIOAction[R, _ <: NoStream, _ <: Effect], actorSystem: ActorSystem): Future[R] = {
+    def retry[T](f: => Future[T], delay: FiniteDuration, retries: Int): Future[T] = {
+      implicit val ec = actorSystem.dispatcher
+      f recoverWith {
+        case e: MySQLTransactionRollbackException if retries > 0 =>
+          /** https://dev.mysql.com/doc/refman/5.5/en/innodb-deadlocks.html
+            *
+            * "Normally, you must write your applications so that they are always
+            *  prepared to re-issue a transaction if it gets rolled back because of a deadlock."
+            */
+          println(s"MySQL Transaction rollback detected.  Retrying in $delay ($retries retries remaining)")
+          after(delay, actorSystem.scheduler)(retry(f, delay, retries - 1))
+      }
+    }
+
+    def queryAttempt = database.run(action.transactionally)
+
+    Future(Await.result(retry(queryAttempt, delay=200 millis, retries=10), Duration.Inf))(actionExecutionContext)
```

Comment:
Good catch Miguel! Sorry Scott. More changes that I missed, again-

In short, I concur that this composite future is inverted.

The whole `f()` to be scheduled for retry should be `DataAccess.upsertWhatever()`. 

If it's easier to see in code, I'll mock up something for discussion as soon as I'm off mobile. I'm roughly picturing a generic method on DataAccess allowing restart on any of the DA methods.